File: scripts/lib/extractors/objc.py
"""Objective-C via tree-sitter-objc.

`@interface` declares a class's API (name, superclass, adopted protocols,
method signatures); `@implementation` carries the method bodies; `@protocol`
declares a protocol. The grammar exposes none of these names through a `name`
field, so we read the first `identifier` child. Call-graph edges come from the
message expressions (`[recv selector]`) in the matching implementation.
"""
from __future__ import annotations
from pathlib import Path
import tree_sitter_objc as _grammar
from tree_sitter import Language, Parser

from .base import Declaration, ImportSpec, ParseResult
from ._common import text_of, walk, loc_of

name = "objc"
extensions = (".m", ".mm")
grammar_package = "tree-sitter-objc"

_LANG = Language(_grammar.language())
_PARSER = Parser(_LANG)


def _first_identifier(node, src):
    for c in node.children:
        if c.type == "identifier":
            return text_of(c, src)
    return None


def _path_to_namespace(rel: str) -> str:
    p = Path(rel)
    parts = list(p.parts[:-1]) + [p.stem]
    while parts and parts[0] in ("src", "Sources", "Classes"):
        parts = parts[1:]
    return ".".join(parts)


def _supertypes(node, src):
    out = []
    for c in node.children:
        if c.type == "superclass_reference":
            ident = next((d for d in c.children if d.type == "identifier"), None)
            if ident is not None:
                out.append(text_of(ident, src))
        elif c.type == "protocol_reference_list":
            for d in walk(c):
                if d.type == "identifier":
                    out.append(text_of(d, src))
    return out


def _method_count(node) -> int:
    body = next((c for c in node.children
                 if c.type in ("interface_declaration_list", "implementation_definition_list")), None)
    if body is None:
        return 0
    return sum(1 for c in body.children
               if c.type in ("method_declaration", "method_definition"))


def _selector_name(msg_node, src):
    """`[self doThing]` → 'doThing'; `[obj run:1 with:2]` → 'run' (first keyword)."""
    sel = msg_node.child_by_field_name("selector")
    if sel is None:
        return None
    if sel.type == "identifier":
        return text_of(sel, src)
    for d in walk(sel):
        if d.type == "identifier":
            return text_of(d, src)
    return None


def _body_refs(node, src) -> list[str]:
    names = set()
    for n in walk(node):
        if n.type == "message_expression":
            nm = _selector_name(n, src)
            if nm:
                names.add(nm)
    return sorted(names)


def parse(path: Path, src: bytes, project_root: Path) -> ParseResult:
    rel = str(path.relative_to(project_root))
    namespace = _path_to_namespace(rel)
    root = _PARSER.parse(src).root_node

    imports = []
    for n in walk(root):
        if n.type in ("preproc_include", "preproc_import"):
            for c in n.children:
                if c.type in ("string_literal", "system_lib_string"):
                    raw = text_of(c, src).strip('"<>')
                    imports.append(ImportSpec(raw=raw, qualified=None, alias=raw.rsplit("/", 1)[-1]))
    imp_refs = [i.qualified for i in imports if i.qualified]

    # Gather implementation bodies first so an @interface node can borrow the
    # call-graph edges from its matching @implementation in the same file.
    impl_refs: dict[str, list[str]] = {}
    impl_nodes: dict[str, object] = {}
    for c in root.children:
        if c.type == "class_implementation":
            cname = _first_identifier(c, src)
            if cname:
                impl_refs[cname] = _body_refs(c, src)
                impl_nodes[cname] = c

    decls, skipped = [], []
    seen_interface = set()
    for c in root.children:
        if c.type == "class_interface":
            cname = _first_identifier(c, src)
            if not cname:
                skipped.append({"path": rel, "line": c.start_point[0]+1, "reason": "no_class_name"})
                continue
            seen_interface.add(cname)
            decls.append(Declaration(
                name=cname, namespace=namespace or None, kind="class",
                path=rel, line=c.start_point[0]+1, supertypes=_supertypes(c, src),
                refs=impl_refs.get(cname, []) + imp_refs, loc=loc_of(c),
                method_count=_method_count(c),
            ))
        elif c.type == "protocol_declaration":
            pname = _first_identifier(c, src)
            if not pname:
                continue
            decls.append(Declaration(
                name=pname, namespace=namespace or None, kind="protocol",
                path=rel, line=c.start_point[0]+1, supertypes=_supertypes(c, src),
                refs=imp_refs, loc=loc_of(c), method_count=_method_count(c),
            ))

    # Implementations with no @interface in this file (e.g. plain .m files).
    for cname, node in impl_nodes.items():
        if cname in seen_interface:
            continue
        decls.append(Declaration(
            name=cname, namespace=namespace or None, kind="class",
            path=rel, line=node.start_point[0]+1, supertypes=_supertypes(node, src),
            refs=impl_refs.get(cname, []) + imp_refs, loc=loc_of(node),
            method_count=_method_count(node),
        ))
    return ParseResult(declarations=decls, imports=imports, skipped=skipped)

File: scripts/lib/extractors/objc.mjs
// scripts/lib/extractors/objc.mjs — Objective-C via tree-sitter-objc.
// Faithful port of scripts/lib/extractors/objc.py.
//
// `@interface` declares a class's API (name, superclass, adopted protocols,
// method signatures); `@implementation` carries the method bodies; `@protocol`
// declares a protocol. The grammar exposes none of these names through a `name`
// field, so we read the first `identifier` child. Call-graph edges come from the
// message expressions (`[recv selector]`) in the matching implementation.
import { init, loadLanguage, Parser } from '../ts.mjs';
import { Declaration, ImportSpec, ParseResult } from './base.mjs';
import { textOf, walk, locOf } from './_common.mjs';

export const name = 'objc';
export const extensions = ['.m', '.mm'];
export const grammar = 'objc'; // manifest key

let _lang;
async function ensure() {
  if (_lang) return;
  await init();
  _lang = await loadLanguage('objc');
}

function firstIdentifier(node, src) {
  for (const c of node.children) {
    if (c.type === 'identifier') return textOf(c, src);
  }
  return null;
}

function pathToNamespace(rel) {
  const segs = rel.split('/');
  const file = segs.pop();
  const stem = file.replace(/\.[^.]+$/, ''); // drop final extension
  let parts = [...segs, stem];
  while (parts.length && (parts[0] === 'src' || parts[0] === 'Sources' || parts[0] === 'Classes')) {
    parts = parts.slice(1);
  }
  return parts.join('.');
}

function supertypesOf(node, src) {
  const out = [];
  let sawColon = false;
  for (const c of node.children) {
    if (c.type === ':') {
      sawColon = true; // the `identifier` right after `:` is the superclass
    } else if (c.type === 'superclass_reference') {
      // Older grammars wrap the superclass.
      let ident = null;
      for (const d of c.children) { if (d.type === 'identifier') { ident = d; break; } }
      if (ident != null) out.push(textOf(ident, src));
    } else if (sawColon && c.type === 'identifier') {
      out.push(textOf(c, src)); // bare superclass identifier (this grammar)
      sawColon = false;
    } else if (c.type === 'protocol_reference_list' || c.type === 'parameterized_arguments') {
      // Adopted protocols: `<P1, P2>`. Identifiers in older grammars,
      // `type_identifier` inside `type_name` in this one.
      for (const d of walk(c)) {
        if (d.type === 'identifier' || d.type === 'type_identifier') out.push(textOf(d, src));
      }
    }
  }
  return out;
}

// Count method declarations/definitions belonging to this @interface/@implementation.
// Grammars vary: some wrap the body in a *_list / implementation_definition node,
// this one lists methods (or single-method wrappers) directly under the decl.
function methodCount(node) {
  const BODY = new Set([
    'interface_declaration_list', 'implementation_definition_list',
    'implementation_definition',
  ]);
  const METHOD = new Set(['method_declaration', 'method_definition']);
  let n = 0;
  for (const c of node.children) {
    if (METHOD.has(c.type)) {
      n++;
    } else if (BODY.has(c.type)) {
      for (const d of c.children) {
        if (METHOD.has(d.type)) n++;
        else if (BODY.has(d.type)) {
          for (const e of d.children) if (METHOD.has(e.type)) n++;
        }
      }
    }
  }
  return n;
}

// `[self doThing]` → 'doThing'; `[obj run:1 with:2]` → 'run' (first keyword).
// This grammar exposes the receiver via a `selector`/`receiver` field but folds
// the selector keywords into bare `identifier` children after the receiver; the
// selector name is the first identifier that is not the receiver.
function selectorName(msgNode, src) {
  // Older grammars expose a dedicated `selector` field.
  const sel = msgNode.childForFieldName('selector');
  if (sel != null) {
    if (sel.type === 'identifier') return textOf(sel, src);
    for (const d of walk(sel)) {
      if (d.type === 'identifier') return textOf(d, src);
    }
    return null;
  }
  // This grammar: identifiers are direct children; skip the receiver.
  const receiver = msgNode.childForFieldName('receiver');
  for (const c of msgNode.children) {
    if (c.type === 'identifier' && c.id !== (receiver && receiver.id)) {
      return textOf(c, src);
    }
  }
  return null;
}

function bodyRefs(node, src) {
  const names = new Set();
  for (const n of walk(node)) {
    if (n.type === 'message_expression') {
      const nm = selectorName(n, src);
      if (nm) names.add(nm);
    }
  }
  return [...names].sort();
}

export async function parse(relPath, src, _projectRoot) {
  await ensure();
  const namespace = pathToNamespace(relPath);
  const parser = new Parser();
  parser.setLanguage(_lang);
  const root = parser.parse(src).rootNode;

  const imports = [];
  for (const n of walk(root)) {
    if (n.type === 'preproc_include' || n.type === 'preproc_import') {
      for (const c of n.children) {
        if (c.type === 'string_literal' || c.type === 'system_lib_string') {
          const raw = textOf(c, src).replace(/^["<>]+|["<>]+$/g, '');
          imports.push(ImportSpec(raw, null, raw.split('/').pop()));
        }
      }
    }
  }
  const impRefs = imports.filter((i) => i.qualified).map((i) => i.qualified);

  // Gather implementation bodies first so an @interface node can borrow the
  // call-graph edges from its matching @implementation in the same file.
  const implRefs = new Map();
  const implNodes = new Map();
  for (const c of root.children) {
    if (c.type === 'class_implementation') {
      const cname = firstIdentifier(c, src);
      if (cname) {
        implRefs.set(cname, bodyRefs(c, src));
        implNodes.set(cname, c);
      }
    }
  }

  const decls = [];
  const skipped = [];
  const seenInterface = new Set();
  for (const c of root.children) {
    if (c.type === 'class_interface') {
      const cname = firstIdentifier(c, src);
      if (!cname) {
        skipped.push({ path: relPath, line: c.startPosition.row + 1, reason: 'no_class_name' });
        continue;
      }
      seenInterface.add(cname);
      decls.push(Declaration({
        name: cname,
        namespace: namespace || null,
        kind: 'class',
        path: relPath,
        line: c.startPosition.row + 1,
        supertypes: supertypesOf(c, src),
        refs: (implRefs.get(cname) ?? []).concat(impRefs),
        loc: locOf(c),
        method_count: methodCount(c),
      }));
    } else if (c.type === 'protocol_declaration') {
      const pname = firstIdentifier(c, src);
      if (!pname) continue;
      decls.push(Declaration({
        name: pname,
        namespace: namespace || null,
        kind: 'protocol',
        path: relPath,
        line: c.startPosition.row + 1,
        supertypes: supertypesOf(c, src),
        refs: impRefs,
        loc: locOf(c),
        method_count: methodCount(c),
      }));
    }
  }

  // Implementations with no @interface in this file (e.g. plain .m files).
  for (const [cname, node] of implNodes) {
    if (seenInterface.has(cname)) continue;
    decls.push(Declaration({
      name: cname,
      namespace: namespace || null,
      kind: 'class',
      path: relPath,
      line: node.startPosition.row + 1,
      supertypes: supertypesOf(node, src),
      refs: (implRefs.get(cname) ?? []).concat(impRefs),
      loc: locOf(node),
      method_count: methodCount(node),
    }));
  }
  return ParseResult(decls, imports, skipped);
}
